### ISSUE TITLE
Technique H71: Better advice on form grouping in practice

### ISSUE THREAD

**Issue by yatil** [w3c/wcag#261]
It was pointed out to me that [Tech H71](https://www.w3.org/TR/WCAG20-TECHS/H71.html) has an example that does not quite work in practice due to screen reader differences:

[Example 4](https://www.w3.org/TR/WCAG20-TECHS/H71.html#H71-ex4) reads like this:

> ### Example 4: Logically related controls
> 
> In this example, form fields for residential and postal addresses are distinguished by the value of the legend in each fieldset grouping.
> 
> Example Code:
> ```html
> <form action="http://example.com/adduser" method="post">
>    <fieldset>
>      <legend>Residential Address</legend>
>      <label for="raddress">Address: </label>
>      <input type="text" id="raddress" name="raddress" />
>      <label for="rzip">Postal/Zip Code: </label>
>      <input type="text" id="rzip" name="rzip" />
>      ...more residential address information...
>    </fieldset>
>    <fieldset>
>      <legend>Postal Address</legend>
>      <label for="paddress">Address: </label>
>      <input type="text" id="paddress" name="paddress" />
>      <label for="pzip">Postal/Zip Code: </label>
>      <input type="text" id="pzip" name="pzip" />
>      ...more postal address information...
>    </fieldset>
> </form>
> ```

Screen readers sometimes do or do not support attaching the text of the legend to the labels. Some attach it to every label. It might also be a preference. However in this example the Residential vs. Postal distinction is _essential_, so authors should not risk that those groupings are missed.

In the [forms tutorial](https://www.w3.org/WAI/tutorials/forms/grouping/#related-fields) we therefore suggest that people put the (abbreviated) legend also into the first label of the group, but hide it from sight. Of course there is also a way to use ARIA but that seems unnecessarily complicated.

My suggestion for the group is to replace the example above with the one in the tutorial.

Maybe we need to reach out to @w3c/aria and see if there is a need for tweaking the accessible name generation algorithm. (I have not looked into that aspect at all.)

[Assigned to @joshueoconnor and @awkawk because you know what to do next :-)]

**Comment by awkawk** [comment]
The technique represents one way to meet an SC, not the only way. Screen readers do have preferences around fieldset/legend, and it is up to the user's preference as to what level of verbosity they want to experience. 

So, I'd say that we can make an additional technique based on the example in the forms technique, but this one is ok.

**Comment by patrickhlauke** [comment]
Screen readers will usually apply (different) heuristics for when to announce the legend. For instance, they may not announce it for every single control in the fieldset, but only do it once the fieldset is entered (so not necessarily only for the first, in DOM order, form control, but for the first form control that receives focus in that fieldset, regardless of how it was reached).

Adding explicit `aria-label`, `aria-describedby` or similar to each form control will *force* screen readers to announce the legend (or abbreviated legend, depending what you do) for each form control, but that may well result in excessive verbosity.

I'm concerned that if the example for the technique went down the path of adding "Postal"/"Residential" explicitly to each label or whatever, readers may misunderstand this to mean they *must* do this, always.

**Comment by yatil** [comment]
@patrickhlauke That’s why no-one suggests to add “"Postal"/"Residential" explicitly to each label”. :-)

**Comment by patrickhlauke** [comment]
adding an extra hidden label for the first field is also a technique that i'd be worried authors will simply cargo-cult. if users disable their AT from announcing legend, is that not user responsibility? users who have legend announced will hear it twice for that field, which is probably only a minor inconvenience.

also, adding it just to the first form control will do nothing for situations where user jumps directly to another form control  (maybe they navigate in reverse, or used a separate AT dialog to find all form controls and jump straight to one of them)?

anyway, just generally concerned, is all ;)

**Comment by yatil** [comment]
Thanks for your input. I agree with your concerns. The technique has a lot of caveats, so has every other.

>  if users disable their AT from announcing legend, is that not user responsibility?

Some AT by default don’t announce it. Should the user know they have to enable it (I think it’s even not an option in some).

**Comment by yatil** [comment]
@awkawk I’m ok with it if you’re (WCAG) is ok with it.

**Comment by joshueoconnor** [comment]
<non sequitur rant>
Personally, while I see why patterns like this have come about, in practice the use of <fieldset>/<legend> sometimes result in the <legend> being announced for each <input> element within the fieldset. It can be incredible annoying.
</non sequitur rant>

**Comment by patrickhlauke** [comment]
one way (when they support it badly) or another (when they don't, or don't do it by default), these seem to be UA/AT issues more than author problems (or rather, authors being asked to code around the failings of current UA/AT). as long as it's framed that way, fine :) (but i'd love to collect cases like this and submit bugs/feature requests to UA/AT vendors so they get their act together...wishful thinking, i know)

**Comment by MakotoUeki** [comment]
I agree with Andrew. This is an "accessibility supported" issue. Techniques documented in "Technique for WCAG 2.0" don't need to work in practice. For example, if it follows HTML spec, it can be a technique. And it doesn't mean that using the technique could meet SC. It depends on how well the technique is supported by targeted UA/AT for the web page. Actually, not all of the techniques work with the Japanese screen readers.

Just FYI. In Japan, we developed a bunch of test files based on "Techniques for WCAG 2.0" about 8 years ago. And we tested them with Japanese UA/ATs to determine if each technique is accessibility-supported. 
Here are examples:
http://waic.jp/docs/as/info/201406/H71-1.html
http://waic.jp/docs/as/info/201406/H71-2.html
http://waic.jp/docs/as/info/201406/H71-3.html
http://waic.jp/docs/as/info/201406/H71-4.html

We are updating the test files. It has been very tough and time-consuming though ;-)

**Comment by awkawk** [comment]
I think that the double voicing is a substantial problem with the example - I don't think that users would want to hear "shipping address. shipping. name." and might even find it confusing. "shipping address. shipping address. name" (where the hidden text matches the legend) is possibly more comprehensible, but still redundant.  I'd rather promote good semantic practice and encourage the tools to do the same. I would hate to have AT tool vendors thinking "we could support fieldset properly but we don't really need to since this design pattern is widespread"...

**Comment by yatil** [comment]
@MakotoUeki, thanks for the research.

@all I am happy to let go of this, I don’t feel strongly and I value your different specific aspects.

**Comment by patrickhlauke** [comment]
Having said that (though that will require revisiting/maintenance) a note after current technique saying that some UAs/ATs don't currently support this as per spec wouldn't harm?

**Comment by awkawk** [comment]
Thank you for volunteering to do that? :)

I'd say that a general note that highlights that support varies and that some tools control the behavior in preferences might be a good way, more detailed information about accessibility support will change more rapidly than we will keep up with.
